analysers/analyser_osmosis_indoor.py

```diff
@@ -0,0 +1,226 @@
+FROM
+  indoor_surfaces
+WHERE
+   NOT tags?'level'
+"""
+
+sql50 = """
+SELECT
+  id,
+  geom
+FROM
+  indoor_surfaces
+WHERE
+  indoor_surfaces.indoor != 'room'
+  AND indoor_surfaces.tags?'shop'
+"""
+
+sql60 = """
+CREATE TEMP TABLE indoor_surfaces_connected_to_highways AS
+(
+SELECT DISTINCT
+  indoor_surfaces.id,
+  indoor_surfaces.geom,
+  indoor_surfaces.indoor,
+  indoor_surfaces.level as surface_level,
+  highway_ends.nid,
+  highway_ends.highway,
+  highways.tags->'level' as connected_highway_level
+FROM
+  indoor_surfaces
+JOIN way_nodes ON indoor_surfaces.id = way_nodes.way_id
+JOIN nodes ON nodes.id = way_nodes.node_id
+JOIN highway_ends ON nodes.id = highway_ends.nid
+JOIN highways ON highway_ends.id = highways.id
+WHERE
+  indoor_surfaces.tags->'indoor' IN ('room', 'corridor', 'area')
+  AND highways.highway IN ('steps', 'footway', 'pedestrian')
+)
+"""
+
+# TODO : check that all surfaces have at least one connected_highway_level that matches it own surface_level
+# (dealing with level="0;1" that occurs on highway=steps
+# and assuming than no level on highway is probably implicit level=0)
+
+sql61 = """
+CREATE TEMP TABLE indoor_surfaces_connected_to_other_surfaces AS
+(
+  SELECT DISTINCT
+  i1.id as id,
+  i1.level as surface_level,
+  i2.id as other_surface_id,
+  i2.level as other_surface_level
+FROM
+  way_nodes w2
+JOIN way_nodes w1 ON w1.node_id = w2.node_id
+JOIN indoor_surfaces i1 ON i1.id = w1.way_id
+JOIN indoor_surfaces i2 ON i2.id = w2.way_id
+WHERE
+  i1.id <> i2.id
+  AND i1.tags->'indoor' IN ('room', 'corridor', 'area')
+)
+""" # maybe check the levels too to make sure they are actually connected ?
+
+sql62 = """
+SELECT
+  indoor_surfaces.id,
+  indoor_surfaces.geom
+FROM
+  indoor_surfaces
+LEFT JOIN indoor_surfaces_connected_to_other_surfaces ON indoor_surfaces_connected_to_other_surfaces.id = indoor_surfaces.id
+LEFT JOIN indoor_surfaces_connected_to_highways ON indoor_surfaces_connected_to_highways.id = indoor_surfaces.id
+WHERE
+  indoor_surfaces.tags->'indoor' IN ('room', 'corridor', 'area')
+  AND indoor_surfaces_connected_to_highways.id IS NULL
+  AND indoor_surfaces_connected_to_other_surfaces.id is NULL
+"""
+
+class Analyser_Osmosis_Indoor(Analyser_Osmosis):
+    requires_tables_common = ['highway_ends']
+
+    def __init__(self, config, logger = None):
+        Analyser_Osmosis.__init__(self, config, logger)
+        self.classs[1] = self.def_class(item = 9999, level = 3, tags = ['indoor', 'geom', 'fix:chair'],
+            title = T_('This indoor feature should be a closed and valid polygon'))
+        self.classs[2] = self.def_class(item = 9999, level = 3, tags = ['indoor', 'geom', 'fix:chair'],
+            title = T_('This indoor feature should be a polygon instead of a point'),
+            fix = T_(
+'''If this feature is actually an indoor area, try to map it as a closed way.
+If this is an indoor object (any kind of feature located inside a building),
+consider using indoor=yes instead'''))
```

Comment:
I've forgotten this often myself, so I know you'll be requested to put periods at the end of sentences in `details` and `fix` items :)